druid/src/scroll_component.rs

```diff
@@ -0,0 +1,456 @@
+    /// the initial x-offset of the dragging input
+    Horizontal(f64),
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct ScrollbarsState {
+    pub opacity: f64,
+    pub timer_id: TimerToken,
+    pub hovered: BarHoveredState,
+    pub held: BarHeldState,
+}
+
+impl Default for ScrollbarsState {
+    fn default() -> Self {
+        Self {
+            opacity: 0.0,
+            timer_id: TimerToken::INVALID,
+            hovered: BarHoveredState::None,
+            held: BarHeldState::None,
+        }
+    }
+}
+
+impl ScrollbarsState {
+    /// true if either scrollbar is currently held down/being dragged
+    pub fn are_held(&self) -> bool {
+        !matches!(self.held, BarHeldState::None)
+    }
+}
+
+/// Embedable component exposing reusable scroll handling logic.
+///
+/// In most situations composing [`Scroll`] or [`List`] is a better idea
+/// for general UI construction. However some cases are not covered by
+/// composing those widgets, such as when a widget needs fine grained
+/// control over its scrolling state or doesn't make sense to exist alone
+/// without scrolling behavior.
+///
+/// `ScrollComponent` contains the unified and consistant scroll logic
+/// used by both [`Scroll`] and [`List`]. This can be used to add this
+/// logic to a custom widget when the need arises.
+///
+/// To use, instance in your widget's new fn and place in a field, keep
+/// the [`content_size`] field updated as scrollable content size changes,
+/// call [`event`] and [`lifecycle`] with all event and lifecycle events,
+/// call [`handle_scroll`] with all events after all other event handling
+/// code and finally perform painting from within a closure provided to
+/// [`paint_content`].
+///
+/// [`Scroll`]: ../widget/struct.Scroll.html
+/// [`List`]: ../widget/struct.List.html
+/// [`content_size`]: struct.ScrollComponent.html#field.content_size
+/// [`event`]: struct.ScrollComponent.html#method.event
+/// [`handle_scroll`]: struct.ScrollComponent.html#method.handle_scroll
+/// [`lifecycle`]: struct.ScrollComponent.html#method.lifecycle
+/// [`paint_content`]: struct.ScrollComponent.html#method.paint_content
+#[derive(Debug, Copy, Clone)]
+pub struct ScrollComponent {
+    /// The size of the scrollable content, make sure to keep up this
+    /// accurate to the content being scrolled
+    pub content_size: Size,
+    pub scroll_offset: Vec2,
+    pub scrollbars: ScrollbarsState,
+}
+
+impl Default for ScrollComponent {
+    fn default() -> Self {
+        ScrollComponent::new()
+    }
+}
+
+impl ScrollComponent {
+    pub fn new() -> ScrollComponent {
+        ScrollComponent {
+            content_size: Default::default(),
+            scroll_offset: Vec2::new(0.0, 0.0),
+            scrollbars: ScrollbarsState::default(),
+        }
+    }
+
+    /// Update the scroll.
+    ///
+    /// Returns `true` if the scroll has been updated.
+    pub fn scroll(&mut self, delta: Vec2, size: Size) -> bool {
```

Comment:
Hmmm good point about the rename though it would be better to name it `layout_size` or something like that as it's not the size of the viewport but instead the scrolling widget's layout size.

I don't think it should be stored in the struct is because we can get the current and correct size whenever this function needs to be called (such as within layout or event) which reduces the amount of paperwork it and the embedding widget needs to do. I hadn't thought about this before but I like it as it is.